### PR TITLE
fix: prevent duplicate 'Resolved' comments and resolve threads

### LINKED ISSUE #5: bug: Multiple duplicate "Resolved" comments being posted
## Description

When marking issues as resolved, the tool posts multiple duplicate "✅ Resolved" comments on the same thread, as shown in the screenshot.

## Current Behavior

Each run of the reviewer posts new "Resolved" replies even when the comment already has a "Resolved" reply from a previous run. This creates spam and noise in the PR.

## Expected Behavior

Only post ONE "Resolved" reply per original comment. Skip if already marked resolved.

## Root Cause

The `_get_resolved_comment_ids()` method checks `comment.in_reply_to_id` to find already-resolved comments, but this may not be working correctly with the GitHub API response structure.

## Files to Modify

- `src/ai_reviewer/github/client.py` - `resolve_fixed_comments()` and `_get_resolved_comment_ids()`

## Acceptance Criteria

- [ ] Only one "Resolved" comment per original finding
- [ ] Subsequent runs skip already-resolved comments
- [ ] Add test coverage for duplicate detection

### PR DESCRIPTION
**Pull request by xilosada**
## Summary

This PR fixes issue #5 with a clean, comprehensive solution.

## Root Cause

The bot was treating its own "✅ **Resolved**" reply comments as new findings, causing an exponential cascade of duplicate comments. Each review cycle would:
1. See the "Resolved" reply (which matches AI reviewer format due to bold text)
2. Treat it as a finding
3. Mark it as "fixed" (since it didn't match any current actual findings)
4. Post another "Resolved" reply

### Core Bug Fix
- **Exclude resolved replies from findings**: `get_previous_review_comments()` now explicitly skips comments containing "✅ **Resolved**"

### Duplicate Detection Improvements
- Safely handle PyGithub's `NotSet`, `None`, and `0` for `in_reply_to_id`
- Filter to only count resolved replies from allowed AI reviewer users
- Handle `None` user and `None` login gracefully

### API Efficiency
- Cache current user login (avoid redundant API calls)
- Use explicit `_USER_FETCH_FAILED` sentinel for failed fetches
- Cache `allowed_users` set
- Fetch review comments once and pass to helpers

### Thread Resolution via GraphQL
- Add GraphQL support for resolving review threads (REST API doesn't support this)
- Fix GitHub Enterprise GraphQL URL construction (`/api/v3` → `/api/graphql`)

### Race Condition Prevention
- Add concurrency control to workflow: only one review per PR at a time

## Testing

- 13 new unit tests covering all the above scenarios
- All tests pass (except pre-existing failure tracked in #25)

## Files Changed

- `src/ai_reviewer/github/client.py` - Core fix and improvements
- `.github/workflows/ai-review.yaml` - Concurrency control
- `tests/test_github.py` - New test cases

Fixes #5

---

> [!NOTE]
> **Medium Risk**
> Touches GitHub review-comment handling and adds GraphQL calls using a PAT, so regressions could affect review posting/resolution behavior across PRs; changes are guarded (caching, max page limit) and covered by new tests.
> 
> **Overview**
> Prevents the bot from re-processing its own `✅ **Resolved**` replies as findings, tightening duplicate-detection logic for already-resolved comments (handles PyGithub `NotSet`/`None`/`0`, filters by allowed users, and avoids redundant comment fetches via caching).
> 
> Adds GraphQL support to actually *resolve* PR review threads (including GitHub Enterprise GraphQL URL handling), and updates the GitHub Actions workflow to serialize runs per PR and prefer `GH_PAT` over `GITHUB_TOKEN` for full thread-resolution capabilities; README documents the PAT requirement and setup.
> 
> Extends unit test coverage around resolved-comment parsing, caching behavior, API-call efficiency, and thread-mapping pagination limits.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a46fb2350ed889acaaacc40540d40ea82c9d9549. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/ai_reviewer/github/client.py [hunk 1]
```diff
@@ -439,6 +481,194 @@ def _normalize_title(self, title: str) -> str:
...
+        owner, name = repo_name.split("/")
+        comment_to_thread: dict[int, str] = {}
+
+        query = """
+        query($owner: String!, $name: String!, $pr_number: Int!, $cursor: String) {
+          repository(owner: $owner, name: $name) {
+            pullRequest(number: $pr_number) {
+              reviewThreads(first: 100, after: $cursor) {
+                pageInfo {
+                  hasNextPage
```
[the remark this answers is not in the document]
Comment (xilosada) [reply src/ai_reviewer/github/client.py]: ✅ **Resolved** - This issue has been addressed in the latest changes.

File: src/ai_reviewer/github/client.py [hunk 2]
```diff
@@ -439,6 +481,194 @@ def _normalize_title(self, title: str) -> str:
...
+
+        Returns:
+            Response data dict or None on error
+        """
+        # Determine GraphQL endpoint
+        if self._base_url:
+            # GitHub Enterprise: /api/v3 -> /api/graphql
+            base = self._base_url.rstrip("/")
+            if base.endswith("/api/v3"):
+                graphql_url = base[:-3] + "/graphql"  # Replace /v3 with /graphql
```
[the remark this answers is not in the document]
Comment (xilosada) [reply src/ai_reviewer/github/client.py]: ✅ **Resolved** - This issue has been addressed in the latest changes.

**Review by xilosada — COMMENTED** [review]
## 🤖 MeroReviewer

**Reviewed by 2 agents** | Quality score: 85% | Review time: 106.5s

### 🟡 Issues pending resolution

✅ **12 Fixed** | 🆕 **6 New**

---

### ✅ Fixed Issues

<details>
<summary>The following issues from previous reviews have been addressed:</summary>

1. ~~N+1 GraphQL API calls when resolving threads~~ (`src/ai_reviewer/github/client.py:721`)
2. ~~Thread comment pagination may miss comments~~ (`src/ai_reviewer/github/client.py:555`)
3. ~~Reference to undefined class attribute AI_REVIEWER_USERS~~ (`src/ai_reviewer/github/client.py:99`)
4. ~~Unbounded GraphQL pagination loop~~ (`src/ai_reviewer/github/client.py:554`)
5. ~~GraphQL URL construction allows arbitrary endpoints for Enterprise~~ (`src/ai_reviewer/github/client.py:493`)
6. ~~Unnecessary try/except around isinstance check~~ (`src/ai_reviewer/github/client.py:756`)
7. ~~Missing pagination for comments within review threads~~ (`src/ai_reviewer/github/client.py:556`)
8. ~~No HTTP session reuse for GraphQL requests~~ (`src/ai_reviewer/github/client.py:499`)
9. ~~Non-thread-safe caching of current user login~~ (`src/ai_reviewer/github/client.py:80`)
10. ~~GraphQL error responses may leak sensitive information in logs~~ (`src/ai_reviewer/github/client.py:519`)
11. ~~GraphQL endpoint URL construction relies on trusted base_url~~ (`src/ai_reviewer/github/client.py:493`)
12. ~~Low Severity~~ (`src/ai_reviewer/github/client.py:590`)

</details>

### 🆕 New Issues

*These issues were found in the latest changes:*

### 💡 Suggestion (4)

#### 1. Documentation recommends broad repository scope for PAT
**File:** `README.md` (line 199-207) | **Consensus:** 1/2 agents

The documentation advises users to create a Classic PAT with full `repo` scope, which grants read/write access to all repositories the user can access (including private ones). While this is necessary for the `resolveReviewThread` GraphQL mutation, users should be warned about the security implications of such broad access, especially in organizational contexts.

**Suggested fix:**
```
Add a security note advising users to: (1) use a dedicated bot/service account for the PAT rather than a personal account, (2) regularly rotate the token, and (3) consider the blast radius if the token is compromised. Example: '> **Security tip**: Create a dedicated service account for this PAT to limit exposure if compromised.'
```

> *Found by: security-agent*

#### 2. Inner comments query lacks pagination
**File:** `src/ai_reviewer/github/client.py` (line 545-555) | **Consensus:** 1/2 agents

The GraphQL query fetches `comments(first: 100)` within each thread but doesn't paginate. If a review thread has more than 100 comments and the target comment is beyond the 100th, `_fetch_thread_mapping` won't include it in the mapping. This would cause `_resolve_thread_for_comment` to silently fail to resolve that thread.

**Suggested fix:**
```
Either add pagination for the inner comments query, or document this limitation. In practice, threads rarely exceed 100 comments, but for correctness consider adding a `hasNextPage` check with a warning log when truncation occurs.
```

> *Found by: performance-agent*

#### 3. GraphQL error responses may leak sensitive information in logs
**File:** `src/ai_reviewer/github/client.py` (line 521-521) | **Consensus:** 1/2 agents

The GraphQL error response is logged directly without sanitization. GitHub's GraphQL API error messages can sometimes contain sensitive information like internal identifiers, rate limit details, or authentication context. While this is logged at WARNING level, it could still end up in log aggregation systems.

**Suggested fix:**
```
Consider sanitizing or truncating error messages before logging, or log only a generic error message while storing full details at DEBUG level: `logger.warning(f"GraphQL request failed with {len(result['errors'])} error(s)")` with `logger.debug(f"GraphQL errors: {result['errors']}")`
```

> *Found by: security-agent*

#### 4. Thread mapping fetched even when no threads need resolution
**File:** `src/ai_reviewer/github/client.py` (line 697-699) | **Consensus:** 1/2 agents

The `_fetch_thread_mapping` GraphQL call is made unconditionally before the loop over `fixed_findings`. However, the loop may post replies but then fail to resolve threads (e.g., comment ID not found in mapping). This is efficient for the common case, but the GraphQL call could be skipped entirely if GraphQL thread resolution isn't needed (e.g., when using GITHUB_TOKEN which can't resolve threads anyway).

**Suggested fix:**
```
Consider making the thread mapping fetch lazy or conditional on whether a PAT with GraphQL permissions is available. Alternatively, this is a minor optimization since the batch fetch is already efficient.
```

> *Found by: performance-agent*

### 📝 Nitpick (2)

#### 1. Token stored as instance attribute could be exposed via serialization
**File:** `src/ai_reviewer/github/client.py` (line 68-71) | **Consensus:** 1/2 agents

The raw token is stored in `self._token` as an instance attribute. While this is a common pattern, if the GitHubClient object is ever serialized (pickle, logging with repr, debugging), the token could be unintentionally exposed. The underscore prefix provides some protection by convention but not enforcement.

**Suggested fix:**
```
This is acceptable for most use cases. For higher security environments, consider using a token provider callback pattern or a dedicated secrets manager integration instead of storing the token directly.
```

> *Found by: security-agent*

#### 2. Converting PaginatedList to list loads all comments into memory
**File:** `src/ai_reviewer/github/client.py` (line 686-688) | **Consensus:** 1/2 agents

The line `raw_comments = list(pr.get_review_comments())` eagerly loads all review comments into memory. For PRs with thousands of review comments, this could consume significant memory. The previous lazy iteration pattern would be more memory-efficient.

**Suggested fix:**
```
This tradeoff is intentional to avoid redundant API calls (as noted in the comment). For most PRs this is fine, but consider adding a comment about the memory tradeoff or using itertools.tee if you need to iterate twice without full materialization.
```

> *Found by: performance-agent*


---

<sub>🤖 Generated by [MeroReviewer](https://github.com/calimero-network/ai-code-reviewer) | Review ID: `review-c9c20e47`</sub>

File: src/ai_reviewer/github/client.py [hunk 3]
```diff
@@ -439,6 +481,194 @@ def _normalize_title(self, title: str) -> str:
...
+    def _graphql_request(self, query: str, variables: dict | None = None) -> dict | None:
+        """Make a GraphQL request to GitHub API.
+
+        Args:
+            query: GraphQL query string
+            variables: Optional query variables
+
+        Returns:
+            Response data dict or None on error
+        """
```
[the remark this answers is not in the document]
Comment (xilosada) [reply src/ai_reviewer/github/client.py]: ✅ **Resolved** - This issue has been addressed in the latest changes.

File: src/ai_reviewer/github/client.py [hunk 4]
```diff
@@ -439,6 +481,194 @@ def _normalize_title(self, title: str) -> str:
...
+                "pr_number": pr_number,
+                "cursor": cursor,
+            }
+
+            data = self._graphql_request(query, variables)
+            if not data:
+                break
+
+            pages_fetched += 1
+            pr_data = data.get("repository", {}).get("pullRequest", {})
```
[the remark this answers is not in the document]
Comment (xilosada) [reply src/ai_reviewer/github/client.py]: ✅ **Resolved** - This issue has been addressed in the latest changes.

File: src/ai_reviewer/github/client.py [hunk 5]
```diff
@@ -439,6 +481,194 @@ def _normalize_title(self, title: str) -> str:
...
+        This batches the GraphQL calls to avoid N+1 queries when resolving multiple
+        threads.
+
+        Args:
+            repo_name: Repository in "owner/name" format
+            pr_number: Pull request number
+
+        Returns:
+            Dict mapping comment database IDs to their thread's GraphQL node ID
+            (only includes unresolved threads)
```
Comment (xilosada) [inline src/ai_reviewer/github/client.py]: 💡 **Inner comments query lacks pagination**

The GraphQL query fetches `comments(first: 100)` within each thread but doesn't paginate. If a review thread has more than 100 comments and the target comment is beyond the 100th, `_fetch_thread_mapping` won't include it in the mapping. This would cause `_resolve_thread_for_comment` to silently fail to resolve that thread.

**Suggested fix:**
```
Either add pagination for the inner comments query, or document this limitation. In practice, threads rarely exceed 100 comments, but for correctness consider adding a `hasNextPage` check with a warning log when truncation occurs.
```

**Review by xilosada — COMMENTED** [review]
## 🤖 MeroReviewer

**Reviewed by 2 agents** | Quality score: 85% | Review time: 337.1s

### 🟡 Issues pending resolution

✅ **17 Fixed** | 🆕 **1 New** | ⏳ **1 Open**

---

### ✅ Fixed Issues

<details>
<summary>The following issues from previous reviews have been addressed:</summary>

1. ~~N+1 GraphQL API calls when resolving threads~~ (`src/ai_reviewer/github/client.py:721`)
2. ~~Thread comment pagination may miss comments~~ (`src/ai_reviewer/github/client.py:555`)
3. ~~Reference to undefined class attribute AI_REVIEWER_USERS~~ (`src/ai_reviewer/github/client.py:99`)
4. ~~Unbounded GraphQL pagination loop~~ (`src/ai_reviewer/github/client.py:554`)
5. ~~GraphQL URL construction allows arbitrary endpoints for Enterprise~~ (`src/ai_reviewer/github/client.py:493`)
6. ~~Unnecessary try/except around isinstance check~~ (`src/ai_reviewer/github/client.py:756`)
7. ~~Missing pagination for comments within review threads~~ (`src/ai_reviewer/github/client.py:556`)
8. ~~No HTTP session reuse for GraphQL requests~~ (`src/ai_reviewer/github/client.py:499`)
9. ~~Non-thread-safe caching of current user login~~ (`src/ai_reviewer/github/client.py:80`)
10. ~~GraphQL endpoint URL construction relies on trusted base_url~~ (`src/ai_reviewer/github/client.py:493`)
11. ~~Low Severity~~ (`src/ai_reviewer/github/client.py:590`)
12. ~~Documentation recommends broad repository scope for PAT~~ (`README.md:199`)
13. ~~Inner comments query lacks pagination~~ (`src/ai_reviewer/github/client.py:545`)
14. ~~GraphQL error responses may leak sensitive information in logs~~ (`src/ai_reviewer/github/client.py:521`)
15. ~~Thread mapping fetched even when no threads need resolution~~ (`src/ai_reviewer/github/client.py:697`)
16. ~~Token stored as instance attribute could be exposed via serialization~~ (`src/ai_reviewer/github/client.py:68`)
17. ~~Converting PaginatedList to list loads all comments into memory~~ (`src/ai_reviewer/github/client.py:686`)

</details>

### 🆕 New Issues

*These issues were found in the latest changes:*

### 📝 Nitpick (1)

#### 1. Token stored as instance attribute
**File:** `src/ai_reviewer/github/client.py` (line 68-72) | **Consensus:** 1/2 agents

The GitHub token is stored as `self._token` instance attribute. While necessary for GraphQL requests, this means the token remains in memory for the lifetime of the GitHubClient object. This is generally acceptable but worth noting for security audits - if the object is accidentally serialized or logged, the token could be exposed.

**Suggested fix:**
```
This is standard practice and acceptable. Just ensure GitHubClient instances are not serialized or passed to logging/debugging utilities that might inspect object attributes.
```

> *Found by: security-agent*


### ⏳ Open Issues

*These issues from previous reviews are still present:*

### 💡 Suggestion (1)

#### 1. GraphQL error responses may leak sensitive information in logs
**File:** `src/ai_reviewer/github/client.py` (line 519-521) | **Consensus:** 1/2 agents

The GraphQL error handling logs the full error response from GitHub's API. While this is useful for debugging, GraphQL error messages can sometimes contain sensitive information such as internal IDs, partial query context, or details about the API structure that could be useful to an attacker with log access.

**Suggested fix:**
```
Consider sanitizing or truncating error messages before logging, or use debug-level logging for full error details: `logger.debug(f"GraphQL errors: {result['errors']}")` with a more generic warning message at the warning level.
```

> *Found by: security-agent*


---

<sub>🤖 Generated by [MeroReviewer](https://github.com/calimero-network/ai-code-reviewer) | Review ID: `review-c02d4e92`</sub>

File: README.md [hunk 6]
```diff
@@ -180,6 +180,46 @@ policy:
...
+- ✅ Posting "Resolved" replies
+- ❌ Resolving review threads (requires PAT)
+
+### Full Features (Classic Personal Access Token)
+
+To enable automatic thread resolution when issues are fixed, use a **Classic PAT** (not Fine-grained):
+
+> ⚠️ **Important**: Fine-grained PATs do NOT support the `resolveReviewThread` GraphQL mutation.
+> You must use a Classic PAT with `repo` scope.
+
```
[the remark this answers is not in the document]
Comment (xilosada) [reply README.md]: ✅ **Resolved** - This issue has been addressed in the latest changes.
